src/backend/auxil/range_alloc.rs

```diff
@@ -62,68 +58,48 @@ where
-            }
-        }
 
-        for i in 0..self.free_ranges.len() {
-            // Input is immediately to the left of an existing empty range.
-            if range.end == self.free_ranges[i].start {
-                // Extend this range
-                self.free_ranges[i].start = range.start;
-                // Merge this into an adjacent range to the left if necessary.
-                if i > 0 && self.free_ranges[i - 1].end == self.free_ranges[i].start {
-                    let r = self.free_ranges.remove(i);
-                    self.free_ranges[i - 1].end = r.end;
-                }
-                return Ok(());
+        // Get insertion position.
+        let i = self.free_ranges.iter()
+            .position(|r| r.start > range.start)
+            .unwrap_or(self.free_ranges.len());
+
+        // Try merging with neighboring ranges in the free list.
+        // Before: |left|-(range)-|right|
+        if i > 0 && range.start == self.free_ranges[i - 1].end {
+            // Merge with |left|.
+            self.free_ranges[i - 1].end = range.end;
+            // Check for possibly merge with |right| also.
+            if i < self.free_ranges.len()
+                && self.free_ranges[i - 1].end == self.free_ranges[i].start
+            {
+                let right = self.free_ranges.remove(i);
+                self.free_ranges[i - 1].end = right.end;
             }
-
-            // Input is immediately to the right of an existing empty range.
-            if range.start == self.free_ranges[i].end {
-                // Extend this range
-                self.free_ranges[i].end = range.end;
-
-                // Merge this into an adjacent range to the right if necessary.
-                if i + 1 != self.free_ranges.len()
-                    && self.free_ranges[i + 1].start == self.free_ranges[i].end
-                {
-                    let r = self.free_ranges.remove(i + 1);
-                    self.free_ranges[i].end = r.end;
-                }
-                return Ok(());
+            return;
+        } else if i < self.free_ranges.len() && range.end == self.free_ranges[i].start {
+            // Merge with |right|.
+            self.free_ranges[i].start = range.start;
+            // Check for possibly merge with |left| also.
+            if i > 0 && self.free_ranges[i].start == self.free_ranges[i - 1].end {
+                let left = self.free_ranges.remove(i - 1);
+                self.free_ranges[i].start = left.start;
             }
+            return;
+        }
 
-            // Input is inbetween two empty ranges.
-            if i + 1 != self.free_ranges.len()
-                && range.start > self.free_ranges[i].end
-                && range.end < self.free_ranges[i + 1].start
-            {
-                self.free_ranges.insert(i + 1, range);
-                return Ok(());
-            }
+        // Debug checks
+        if i > 0 {
+            assert!(self.free_ranges[i - 1].end < range.start);
```

Comment:
any reason not to put `i>0` into the assert expression?

Reply:
So `i==0 || ..`? Yeah that could work.